python/ray/data/datasource/parquet_fast_datasource.py

```diff
@@ -0,0 +1,91 @@
+import logging
+from typing import (
+    Dict,
+    List,
+    Tuple,
+    Any,
+    Callable,
+    Optional,
+    TYPE_CHECKING,
+)
+
+if TYPE_CHECKING:
+    import pyarrow
+
+from ray.data.block import BlockAccessor
+from ray.data.datasource.file_based_datasource import (
+    DefaultFileMetadataProvider,
+    FileBasedDatasource,
+    _resolve_kwargs,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class FastFileMetadataProvider(DefaultFileMetadataProvider):
+    """Fast Metadata provider for FileBasedDatasource implementations.
+
+    Offers improved performance vs. DefaultFileMetadataProvider by skipping directory
+    path expansion and file size collection. While this performance improvement may be
+    negligible for local filesystems, it can be substantial for cloud storage service
+    providers.
+
+    This should only be used when all input paths are known to be files.
+    """
+
+    def expand_paths(
+        self,
+        paths: List[str],
+        filesystem: "pyarrow.fs.FileSystem",
+    ) -> Tuple[List[str], List[Optional[int]]]:
+        logger.warning(
+            f"Skipping expansion of {len(paths)} path(s). If your paths contain "
+            f"directories or if file size collection is required, try rerunning this "
+            f"read with `meta_provider=DefaultFileMetadataProvider()`."
+        )
+        import numpy as np
+
+        return paths, np.empty(len(paths), dtype=object)
+
+
+class ParquetFastDatasource(FileBasedDatasource):
```

Comment:
I had some similar thoughts. However, the class itself supports directories, so I'm not sure about either `Directoryless` or `DirectFile`. It's just the `FastFileMetadataProvider` default in the `read_parquet_fast()` API that alters it to not handle directories.

`ParquetNoMetaDatasource` was one thing I was thinking of, since not overriding `prepare_read()` means it's not populated with verbose metadata, but I can also imagine this changing in the future.

I think it may be best to just call this something like `ParquetBasicDatasource` to indicate that it is a minimal implementation that reuses as much code as possible from `FileBasedDatasource`. 

As far as `read_parquet_fast()` API naming goes, I've been thinking that `read_parquet_bulk()` would be more accurate, since the current implementation offers little to no performance benefits when reading a small number of files (esp. if each file is relatively large), but I expect it to continue to offer long-term benefits for bulk reads of a large number of files.

Reply:
My last commit changes `read_parquet_fast()` to `read_parquet_bulk()`, `ParquetFastDatasource` to `ParquetBaseDatasource`, and then has `ParquetDatasource` inherit from `ParquetBaseDatasource` since they share the same file format and write path.

What do you think?

Reply:
Looks good.